src/Protocol.py

```diff
@@ -0,0 +1,289 @@
+            available_keys: set[str] = self.possible_emoji_keys - self.used_keys
+            if len(available_keys) == 0:
+                raise ValueError("No available emoji keys left to assign. Please specify a key for the token.")
+            key = available_keys.pop()
+            token.key = key
+
+        self.tokens.add(token)
+        self.used_keys.add(token.key)
+
+    def add_instruction(self, instruction: Instruction):
+        """
+        Adds an Instruction (and its components) to the protocol.
+
+        Asserts that all samples in the instruction match the defined sample line size.
+        """
+        # Assert all samples match the defined sample line size
+        for sample in instruction.samples:
+            assert len(sample['strings']) == self.instruction_sample_lines, \
+                "The number of sample lines does not match the memory size."
+
+        # Add all token combos as special tokens
+        for token_set in instruction.get_token_sets():
+            self.instruction_token_key_sets.add(token_set.get_token_key_set())
+
+        # Add the result token as a special token
+        if instruction.final.key is not None:
+            self.instruction_token_key_sets.add(instruction.final.key)
+
+        # Add the instruction to the protocol
+        self.instructions.add(instruction)
+
+    def add_number(self, num, min_value, max_value):
+        """Adds a number range to the protocol."""
+        if self.numbers is None:
+            self.numbers = {"None": ''}
+        key = num.value
+        value = f"<Number between {min_value} and {max_value}>"
+        self.numbers[key] = value
+
+    def add_number_list(self, num, min_value, max_value, length):
+        """Adds a number list range to the protocol."""
+        if self.numbers is None:
+            self.numbers = {"None": ''}
+        key = num.value
+        value = f"<List of length {length} of numbers between {min_value} and {max_value}>"
+        self.numbers[key] = value
+
+    def save(self, name: str | None = None, path: str | None = None):
+        """
+        Saves the protocol to a JSON file. This file can be submitted to Databiomes for model training.
+
+        :param name: The name of the file (without extension). If None, uses the protocol's name.
+        :param path: The directory path where the file will be saved. If None, saves in the current directory.
+        """
+        if name is None:
+            name = self.name
+        if path is None:
+            path = os.getcwd()
+        os.makedirs(path, exist_ok=True)
+        filename = f"{path}\\{name}_model.json"
+        print(f"Saving Model Train Protocol to {filename}...")
+        mtp_template = self._serialize()
+        mtp_template = self._rename_template_elements(mtp_template)
+        with open(filename, 'w', encoding="utf-8") as file:
+            json.dump(mtp_template, file, indent=4, ensure_ascii=False)
+
+    def template(self, path: str | None = None):
```

Comment:
When testing template function with example.py I found the following issues:
- No items in index 2 of the input
- Duplicate items in each index (a unique set/example should only appear once in each index of the input if they are found to be used in that index)